### PR TITLE
BUG: TypeError when shifting DataFrame created by concatenation of slices and fills with values

### LINKED ISSUE #42401: BUG: TypeError when shifting DataFrame created by concatenation of slices and fills with values
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import numpy as np
import pandas as pd
npa = np.random.RandomState(0).randint(1000, size=(20, 8))
df = pd.DataFrame(npa, columns=[f'c{i}' for i in range(8)])
pd.concat([df.iloc[:6, 3:5], df.iloc[:6, 6:8]], axis=1).shift(periods=2, axis=1, freq=None, fill_value=0)
```

```
Traceback (most recent call last):
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-9916dba4e800>", line 1, in <module>
    pd.concat([df.iloc[:6, 3:5], df.iloc[:6, 6:8]], axis=1).shift(periods=2, axis=1, freq=None, fill_value=0)
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/frame.py", line 5313, in shift
    return super().shift(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/generic.py", line 9179, in shift
    new_data = self._mgr.shift(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 393, in shift
    result = self.reindex_indexer(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 669, in reindex_indexer
    new_blocks = self._slice_take_blocks_ax0(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 714, in _slice_take_blocks_ax0
    sl_type, slobj, sllen = _preprocess_slice_or_indexer(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 2033, in _preprocess_slice_or_indexer
    raise TypeError(type(slice_or_indexer), dtype)
TypeError: (<class 'list'>, None)
```

#### Problem description

`shift` should work without exception. When `df._data._consolidate_inplace()` is called before shift, it works well. Thus consolidation might be missing, just like #35488.

#### Expected Output

```
   c3  c4   c6   c7
0   0   0  192  835
1   0   0  754  804
2   0   0  705  486
3   0   0  537  845
4   0   0  709  847
5   0   0  755  797
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f00ed8f47020034e752baf0250483053340971b0
python           : 3.8.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Thu May  6 00:48:39 PDT 2021; root:xnu-6153.141.33~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : zh_CN.UTF-8
pandas           : 1.3.0
numpy            : 1.18.5
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.2.2
setuptools       : 49.6.0.post20200814
Cython           : 0.29.23
pytest           : 6.2.2
hypothesis       : 6.3.0
sphinx           : 3.2.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.17.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : 0.4.2
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 4.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.6.0
sqlalchemy       : 1.3.18
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.53.0

</details>

### LINKED ISSUE #42719: BUG: TypeError when shifting DataFrame created by concatenation of slices and fills with values
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import numpy as np
import pandas as pd
npa = np.random.RandomState(0).randint(1000, size=(20, 8))
df = pd.DataFrame(npa, columns=[f'c{i}' for i in range(8)])
pd.concat([df.iloc[:6, 3:5], df.iloc[:6, 6:8]], axis=1).shift(periods=2, axis=1, freq=None, fill_value=0)
```

```
Traceback (most recent call last):
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-9916dba4e800>", line 1, in <module>
    pd.concat([df.iloc[:6, 3:5], df.iloc[:6, 6:8]], axis=1).shift(periods=2, axis=1, freq=None, fill_value=0)
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/frame.py", line 5313, in shift
    return super().shift(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/generic.py", line 9179, in shift
    new_data = self._mgr.shift(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 393, in shift
    result = self.reindex_indexer(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 669, in reindex_indexer
    new_blocks = self._slice_take_blocks_ax0(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 714, in _slice_take_blocks_ax0
    sl_type, slobj, sllen = _preprocess_slice_or_indexer(
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 2033, in _preprocess_slice_or_indexer
    raise TypeError(type(slice_or_indexer), dtype)
TypeError: (<class 'list'>, None)
```

#### Problem description

`shift` should work without exception. When `df._data._consolidate_inplace()` is called before shift, it works well.

I've submitted an issue at #42401 once v1.3.0 is released, but no one responds till v1.3.1 is released today.

#### Expected Output

```
   c3  c4   c6   c7
0   0   0  192  835
1   0   0  754  804
2   0   0  705  486
3   0   0  537  845
4   0   0  709  847
5   0   0  755  797
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : daec2e73b48fe34086dcdeeb0858d29536b6ca6a
python           : 3.8.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.6.0
Version          : Darwin Kernel Version 20.6.0: Wed Jun 23 00:26:31 PDT 2021; root:xnu-7195.141.2~5/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : zh_CN.UTF-8
LOCALE           : zh_CN.UTF-8

pandas           : 1.4.0.dev0+292.gdaec2e73b4
numpy            : 1.19.2
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.3
setuptools       : 52.0.0.post20210125
Cython           : 0.29.24
pytest           : 6.2.2
hypothesis       : 6.3.0
sphinx           : 3.5.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.22.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : 2.7.2
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 4.0.1
pyxlsb           : None
s3fs             : None
scipy            : 1.6.1
sqlalchemy       : 1.4.7
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.51.2

</details>

### PR DESCRIPTION
**Pull request by wjsi**
- [x] closes #42719
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by wjsi** [comment]
Close as cannot handle different managers now.

**Comment by wjsi** [comment]
I decide to reopen this PR and ignore array manager with reference to `test_shift_dt64values_int_fill_deprecated`.

**Comment by jreback** [comment]
looks fine.

cc @jbrockmendel

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -386,10 +386,12 @@ def shift(self: T, periods: int, axis: int, fill_value) -> T:
...
             ncols = self.shape[0]
             if periods > 0:
-                indexer = [-1] * periods + list(range(ncols - periods))
+                indexer = np.concatenate(
+                    [np.repeat(-1, periods), np.arange(ncols - periods)]
+                )
             else:
                 nper = abs(periods)
-                indexer = list(range(nper, ncols)) + [-1] * nper
+                indexer = np.concatenate([np.arange(nper, ncols), np.repeat(-1, nper)])
```
Comment (jbrockmendel) [inline pandas/core/internals/managers.py:394]: i think simpler to just do `np.array(whats_currently_there, dtype=np.intp)`?

**Comment by wjsi** [comment]
@jbrockmendel Done.

File: doc/source/whatsnew/v1.3.2.rst [hunk 2]
```diff
@@ -23,7 +23,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
--
+- Fixed bug in :meth:`DataFrame.shift` where TypeError occurred when shifting DataFrame created by concatenation of slices and fills with values (:issue:`42719`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.3.2.rst:26]: can you move to regression section above (may need to merge master first to avoid conflicts)

Comment (wjsi) [reply doc/source/whatsnew/v1.3.2.rst]: Done.

**Comment by jreback** [comment]
thanks @wjsi

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**